File: src/lexer/lexer_test.zig
const std = @import("std");
const testing = std.testing;
const Lexer = @import("lexer.zig").Lexer;
const TokenType = @import("token.zig").TokenType;
const LexerError = @import("lexer.zig").LexerError;

test "empty input" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 1), tokens.len);
    try testing.expectEqual(TokenType.eof, tokens[0].type);
}

test "whitespace" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, " \t\r\n");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 2), tokens.len);
    try testing.expectEqual(TokenType.newline, tokens[0].type);
    try testing.expectEqual(TokenType.eof, tokens[1].type);
}

test "keywords" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "program when consider otherwise end");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 6), tokens.len);
    try testing.expectEqual(TokenType.program, tokens[0].type);
    try testing.expectEqual(TokenType.when, tokens[1].type);
    try testing.expectEqual(TokenType.consider, tokens[2].type);
    try testing.expectEqual(TokenType.otherwise, tokens[3].type);
    try testing.expectEqual(TokenType.end, tokens[4].type);
    try testing.expectEqual(TokenType.eof, tokens[5].type);
}

test "numbers" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "123 45.67");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 3), tokens.len);
    try testing.expectEqual(TokenType.number, tokens[0].type);
    try testing.expectEqualStrings("123", tokens[0].lexeme);
    try testing.expectEqual(TokenType.number, tokens[1].type);
    try testing.expectEqualStrings("45.67", tokens[1].lexeme);
    try testing.expectEqual(TokenType.eof, tokens[2].type);
}

test "text" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "\"Hello, World!\"");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 2), tokens.len);
    try testing.expectEqual(TokenType.text, tokens[0].type);
    if (tokens[0].literal) |literal| {
        try testing.expectEqualStrings("Hello, World!", literal);
    } else {
        try testing.expect(false);
    }
    try testing.expectEqual(TokenType.eof, tokens[1].type);
}

test "operators" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "+ - * / % = == != < <= > >= && || !");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 16), tokens.len);
    try testing.expectEqual(TokenType.plus, tokens[0].type);
    try testing.expectEqual(TokenType.minus, tokens[1].type);
    try testing.expectEqual(TokenType.multiply, tokens[2].type);
    try testing.expectEqual(TokenType.divide, tokens[3].type);
    try testing.expectEqual(TokenType.modulo, tokens[4].type);
    try testing.expectEqual(TokenType.assign, tokens[5].type);
    try testing.expectEqual(TokenType.equal_to, tokens[6].type);
    try testing.expectEqual(TokenType.not_equal_to, tokens[7].type);
    try testing.expectEqual(TokenType.less_than, tokens[8].type);
    try testing.expectEqual(TokenType.less_equal_to, tokens[9].type);
    try testing.expectEqual(TokenType.greater_than, tokens[10].type);
    try testing.expectEqual(TokenType.greater_equal_to, tokens[11].type);
    try testing.expectEqual(TokenType.and_, tokens[12].type);
    try testing.expectEqual(TokenType.or_, tokens[13].type);
    try testing.expectEqual(TokenType.not_, tokens[14].type);
    try testing.expectEqual(TokenType.eof, tokens[15].type);
}

test "delimiters" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "(){}[],.;:?");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 12), tokens.len);
    try testing.expectEqual(TokenType.left_paren, tokens[0].type);
    try testing.expectEqual(TokenType.right_paren, tokens[1].type);
    try testing.expectEqual(TokenType.left_brace, tokens[2].type);
    try testing.expectEqual(TokenType.right_brace, tokens[3].type);
    try testing.expectEqual(TokenType.left_bracket, tokens[4].type);
    try testing.expectEqual(TokenType.right_bracket, tokens[5].type);
    try testing.expectEqual(TokenType.comma, tokens[6].type);
    try testing.expectEqual(TokenType.dot, tokens[7].type);
    try testing.expectEqual(TokenType.semicolon, tokens[8].type);
    try testing.expectEqual(TokenType.colon, tokens[9].type);
    try testing.expectEqual(TokenType.question_mark, tokens[10].type);
    try testing.expectEqual(TokenType.eof, tokens[11].type);
}

test "line and column tracking" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "x = 42\n  y = 123");
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 8), tokens.len);
    try testing.expectEqual(@as(usize, 1), tokens[0].line); // x
    try testing.expectEqual(@as(usize, 1), tokens[0].column);
    try testing.expectEqual(@as(usize, 1), tokens[1].line); // =
    try testing.expectEqual(@as(usize, 3), tokens[1].column);
    try testing.expectEqual(@as(usize, 1), tokens[2].line); // 42
    try testing.expectEqual(@as(usize, 5), tokens[2].column);
    try testing.expectEqual(@as(usize, 1), tokens[3].line); // newline
    try testing.expectEqual(@as(usize, 7), tokens[3].column);
    try testing.expectEqual(@as(usize, 2), tokens[4].line); // y
    try testing.expectEqual(@as(usize, 3), tokens[4].column);
    try testing.expectEqual(@as(usize, 2), tokens[5].line); // =
    try testing.expectEqual(@as(usize, 5), tokens[5].column);
    try testing.expectEqual(@as(usize, 2), tokens[6].line); // 123
    try testing.expectEqual(@as(usize, 7), tokens[6].column);
    try testing.expectEqual(@as(usize, 2), tokens[7].line); // eof
    try testing.expectEqual(@as(usize, 10), tokens[7].column);
}

test "error handling" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "\"unterminated string");
    defer lexer.deinit();

    _ = lexer.scanTokens() catch |err| {
        try testing.expectEqual(err, LexerError.UnterminatedString);
        try testing.expect(lexer.had_error);
        return;
    };
    try testing.expect(false);
}

test "complex example" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, 
        \\program example
        \\    when x > 10 then
        \\        y = "Hello, World!"
        \\    end
    );
    defer lexer.deinit();

    const tokens = try lexer.scanTokens();
    try testing.expectEqual(@as(usize, 15), tokens.len);
    try testing.expectEqual(TokenType.program, tokens[0].type);
    try testing.expectEqual(TokenType.identifier, tokens[1].type);
    try testing.expectEqual(TokenType.newline, tokens[2].type);
    try testing.expectEqual(TokenType.when, tokens[3].type);
    try testing.expectEqual(TokenType.identifier, tokens[4].type);
    try testing.expectEqual(TokenType.greater_than, tokens[5].type);
    try testing.expectEqual(TokenType.number, tokens[6].type);
    try testing.expectEqual(TokenType.then, tokens[7].type);
    try testing.expectEqual(TokenType.newline, tokens[8].type);
    try testing.expectEqual(TokenType.identifier, tokens[9].type);
    try testing.expectEqual(TokenType.assign, tokens[10].type);
    try testing.expectEqual(TokenType.text, tokens[11].type);
    try testing.expectEqual(TokenType.newline, tokens[12].type);
    try testing.expectEqual(TokenType.end, tokens[13].type);
    try testing.expectEqual(TokenType.eof, tokens[14].type);
}

test "special literals" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "var date = @\"2023-01-15\";\nvar money = @\"$123.45\";");
    defer lexer.deinit();
    
    const tokens = try lexer.scanTokens();
    
    try testing.expect(tokens.len > 0);
    try testing.expectEqual(TokenType.var_, tokens[0].type);
    try testing.expectEqual(TokenType.identifier, tokens[1].type);
    try testing.expectEqualStrings("date", tokens[1].lexeme);
    try testing.expectEqual(TokenType.assign, tokens[2].type);
    try testing.expectEqual(TokenType.special_literal, tokens[3].type);
    try testing.expectEqualStrings("2023-01-15", tokens[3].literal.?);
    try testing.expectEqual(TokenType.semicolon, tokens[4].type);
    try testing.expectEqual(TokenType.newline, tokens[5].type);
    try testing.expectEqual(TokenType.var_, tokens[6].type);
    try testing.expectEqual(TokenType.identifier, tokens[7].type);
    try testing.expectEqualStrings("money", tokens[7].lexeme);
    try testing.expectEqual(TokenType.assign, tokens[8].type);
    try testing.expectEqual(TokenType.special_literal, tokens[9].type);
    try testing.expectEqualStrings("$123.45", tokens[9].literal.?);
    try testing.expectEqual(TokenType.semicolon, tokens[10].type);
}

test "mbl keywords and types" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "var x = true;\nDate myDate = @\"2023-01-15\";\nMoney balance = @\"$100.00\";");
    defer lexer.deinit();
    
    const tokens = try lexer.scanTokens();
    
    try testing.expect(tokens.len > 0);
    try testing.expectEqual(TokenType.var_, tokens[0].type);
    try testing.expectEqual(TokenType.identifier, tokens[1].type);
    try testing.expectEqual(TokenType.assign, tokens[2].type);
    try testing.expectEqual(TokenType.true_, tokens[3].type);
    try testing.expectEqual(TokenType.semicolon, tokens[4].type);
    try testing.expectEqual(TokenType.newline, tokens[5].type);
    try testing.expectEqual(TokenType.date, tokens[6].type);
}

test "mbl comments" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, "var x = 42; // This is a comment\nvar y = 100;");
    defer lexer.deinit();
    
    const tokens = try lexer.scanTokens();
    
    try testing.expect(tokens.len > 0);
    try testing.expectEqual(TokenType.var_, tokens[0].type);
    try testing.expectEqual(TokenType.identifier, tokens[1].type);
    try testing.expectEqual(TokenType.assign, tokens[2].type);
    try testing.expectEqual(TokenType.number, tokens[3].type);
    try testing.expectEqual(TokenType.semicolon, tokens[4].type);
    // The comment should be skipped
    try testing.expectEqual(TokenType.newline, tokens[5].type);
    try testing.expectEqual(TokenType.var_, tokens[6].type);
}

test "mbl trigger syntax" {
    const allocator = testing.allocator;
    var lexer = Lexer.init(allocator, 
        \\trigger balance_changed on data_changed
        \\  when context.key == "balance" then
        \\    log(context.value)
        \\  end
    );
    defer lexer.deinit();
    
    const tokens = try lexer.scanTokens();
    
    try testing.expect(tokens.len > 0);
    try testing.expectEqual(TokenType.trigger, tokens[0].type);
    try testing.expectEqual(TokenType.identifier, tokens[1].type);
    try testing.expectEqualStrings("balance_changed", tokens[1].lexeme);
    try testing.expectEqual(TokenType.on, tokens[2].type);
    try testing.expectEqual(TokenType.identifier, tokens[3].type);
    try testing.expectEqualStrings("data_changed", tokens[3].lexeme);
} 

File: tests/lexer/lexer_test.cpp
#include "lexer/lexer.hpp"
#include "lexer/token.hpp"
#include <gtest/gtest.h>
#include <vector>
#include <chrono>
#include <random>

using namespace mbl;

class LexerTest : public ::testing::Test {
protected:
    std::vector<Token> scan(std::string_view source) {
        Lexer lexer(source);
        return lexer.scanTokens();
    }

    void expectTokens(const std::vector<Token>& tokens, const std::vector<TokenType> expected) {
        ASSERT_EQ(tokens.size(), expected.size());
        for (size_t i = 0; i < tokens.size(); i++) {
            EXPECT_EQ(tokens[i].type(), expected[i]);
        }
    }

    void expectToken(const Token& token, TokenType type, std::string_view lexeme, int line, int column) {
        EXPECT_EQ(token.type(), type);
        EXPECT_EQ(token.lexeme(), lexeme);
        EXPECT_EQ(token.line(), line);
        EXPECT_EQ(token.column(), column);
    }
};

// Basic Input Tests
TEST_F(LexerTest, EmptyInput) {
    auto tokens = scan("");
    expectTokens(tokens, {TokenType::END_OF_FILE});
    expectToken(tokens[0], TokenType::END_OF_FILE, "", 1, 1);
}

TEST_F(LexerTest, Whitespace) {
    auto tokens = scan("   \t\r\n");
    expectTokens(tokens, {TokenType::NEWLINE, TokenType::END_OF_FILE});
    expectToken(tokens[0], TokenType::NEWLINE, "\n", 1, 5);
    expectToken(tokens[1], TokenType::END_OF_FILE, "", 2, 1);
}

// Keyword Tests with Lexeme Verification
TEST_F(LexerTest, Keywords) {
    auto tokens = scan("if then else while do for in function return true false unknown");
    expectTokens(tokens, {
        TokenType::IF,
        TokenType::THEN,
        TokenType::ELSE,
        TokenType::WHILE,
        TokenType::DO,
        TokenType::FOR,
        TokenType::IN,
        TokenType::FUNCTION,
        TokenType::RETURN,
        TokenType::BOOLEAN,
        TokenType::BOOLEAN,
        TokenType::UNKNOWN,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::IF, "if", 1, 1);
    expectToken(tokens[1], TokenType::THEN, "then", 1, 4);
    expectToken(tokens[2], TokenType::ELSE, "else", 1, 9);
    expectToken(tokens[3], TokenType::WHILE, "while", 1, 14);
}

// Number Tests with Lexeme Verification
TEST_F(LexerTest, Numbers) {
    auto tokens = scan("123 456.789 1_234_567");
    expectTokens(tokens, {
        TokenType::NUMBER,
        TokenType::NUMBER,
        TokenType::NUMBER,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::NUMBER, "123", 1, 1);
    expectToken(tokens[1], TokenType::NUMBER, "456.789", 1, 5);
    expectToken(tokens[2], TokenType::NUMBER, "1_234_567", 1, 13);
}

// Text Tests with Line/Column Tracking
TEST_F(LexerTest, Text) {
    auto tokens = scan("\"Hello, World!\" \"Multi\nline\"");
    expectTokens(tokens, {
        TokenType::TEXT,
        TokenType::TEXT,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::TEXT, "\"Hello, World!\"", 1, 1);
    expectToken(tokens[1], TokenType::TEXT, "\"Multi\nline\"", 1, 16);
}

// Special Types with Position Tracking
TEST_F(LexerTest, SpecialTypes) {
    auto tokens = scan("@2024-03-14 $123.45");
    expectTokens(tokens, {
        TokenType::TIME,
        TokenType::MONEY,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::TIME, "@2024-03-14", 1, 1);
    expectToken(tokens[1], TokenType::MONEY, "$123.45", 1, 13);
}

// Comments with Line/Column Tracking
TEST_F(LexerTest, Comments) {
    auto tokens = scan("# Single line comment #\n## Multi\nline\ncomment ##");
    expectTokens(tokens, {
        TokenType::COMMENT_START,
        TokenType::COMMENT_END,
        TokenType::NEWLINE,
        TokenType::COMMENT_START,
        TokenType::COMMENT_END,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::COMMENT_START, "#", 1, 1);
    expectToken(tokens[1], TokenType::COMMENT_END, "#", 1, 21);
    expectToken(tokens[2], TokenType::NEWLINE, "\n", 1, 22);
}

// Edge Cases
TEST_F(LexerTest, EdgeCases) {
    // Test empty comments
    auto tokens1 = scan("##");
    expectTokens(tokens1, {TokenType::COMMENT_START, TokenType::COMMENT_END, TokenType::END_OF_FILE});
    
    // Test unterminated comment
    auto tokens2 = scan("### Comment");
    expectTokens(tokens2, {TokenType::COMMENT_START, TokenType::ERROR, TokenType::END_OF_FILE});
    
    // Test mixed text and operators
    auto tokens3 = scan("x = 42\ny = \"Hello\"\nz > 10");
    expectTokens(tokens3, {
        TokenType::TEXT,
        TokenType::ASSIGN,
        TokenType::NUMBER,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::ASSIGN,
        TokenType::TEXT,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::GREATER,
        TokenType::NUMBER,
        TokenType::END_OF_FILE
    });
}

// Complex Examples with Full Verification
TEST_F(LexerTest, ComplexExample) {
    auto tokens = scan(R"(
if x > 10 then
    y = "Hello"
    computer.write(y)
else
    z = 42
end
)");
    expectTokens(tokens, {
        TokenType::IF,
        TokenType::TEXT,
        TokenType::GREATER,
        TokenType::NUMBER,
        TokenType::THEN,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::ASSIGN,
        TokenType::TEXT,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::DOT,
        TokenType::TEXT,
        TokenType::LEFT_PAREN,
        TokenType::TEXT,
        TokenType::RIGHT_PAREN,
        TokenType::NEWLINE,
        TokenType::ELSE,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::ASSIGN,
        TokenType::NUMBER,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::NEWLINE,
        TokenType::END_OF_FILE
    });
    
    // Verify specific tokens
    expectToken(tokens[0], TokenType::IF, "if", 2, 1);
    expectToken(tokens[1], TokenType::TEXT, "x", 2, 4);
    expectToken(tokens[2], TokenType::GREATER, ">", 2, 6);
    expectToken(tokens[3], TokenType::NUMBER, "10", 2, 8);
}

// Line/Column Tracking Tests
TEST_F(LexerTest, LineColumnTracking) {
    auto tokens = scan(R"(
x = 42
y = "Hello"
z = @2024-03-14
)");
    
    expectToken(tokens[0], TokenType::TEXT, "x", 2, 1);
    expectToken(tokens[1], TokenType::ASSIGN, "=", 2, 3);
    expectToken(tokens[2], TokenType::NUMBER, "42", 2, 5);
    expectToken(tokens[3], TokenType::NEWLINE, "\n", 2, 7);
    
    expectToken(tokens[4], TokenType::TEXT, "y", 3, 1);
    expectToken(tokens[5], TokenType::ASSIGN, "=", 3, 3);
    expectToken(tokens[6], TokenType::TEXT, "\"Hello\"", 3, 5);
    expectToken(tokens[7], TokenType::NEWLINE, "\n", 3, 12);
    
    expectToken(tokens[8], TokenType::TEXT, "z", 4, 1);
    expectToken(tokens[9], TokenType::ASSIGN, "=", 4, 3);
    expectToken(tokens[10], TokenType::TIME, "@2024-03-14", 4, 5);
}

// Recursive Comments Test
TEST_F(LexerTest, RecursiveComments) {
    auto tokens = scan(R"(
# This is a comment #
# Another comment with # inside #
)");
    expectTokens(tokens, {
        TokenType::NEWLINE,
        TokenType::COMMENT_START,
        TokenType::COMMENT_END,
        TokenType::NEWLINE,
        TokenType::COMMENT_START,
        TokenType::COMMENT_END,
        TokenType::NEWLINE,
        TokenType::END_OF_FILE
    });
}

// Multi-Quote Strings Test
TEST_F(LexerTest, MultiQuoteStrings) {
    auto tokens = scan(R"(
"This is a \"quoted\" string"
"String with \"nested\" quotes"
)");
    expectTokens(tokens, {
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::NEWLINE,
        TokenType::TEXT,
        TokenType::NEWLINE,
        TokenType::END_OF_FILE
    });
}

// Money with Currency Test
TEST_F(LexerTest, MoneyWithCurrency) {
    auto tokens = scan("$123.45USD $1,234.56EUR");
    expectTokens(tokens, {
        TokenType::MONEY,
        TokenType::MONEY,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::MONEY, "$123.45USD", 1, 1);
    expectToken(tokens[1], TokenType::MONEY, "$1,234.56EUR", 1, 11);
}

// Time Formats Test
TEST_F(LexerTest, TimeFormats) {
    auto tokens = scan("@09:30 @14:45:30 @2024-03-14T15:30:00");
    expectTokens(tokens, {
        TokenType::TIME,
        TokenType::TIME,
        TokenType::TIME,
        TokenType::END_OF_FILE
    });
    
    expectToken(tokens[0], TokenType::TIME, "@09:30", 1, 1);
    expectToken(tokens[1], TokenType::TIME, "@14:45:30", 1, 8);
    expectToken(tokens[2], TokenType::TIME, "@2024-03-14T15:30:00", 1, 17);
}

// Large Input Performance Test
TEST_F(LexerTest, LargeInputPerformance) {
    std::string input;
    for (int i = 0; i < 10000; i++) {
        input += "x = 42\n";
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    auto tokens = scan(input);
    auto end = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    EXPECT_LT(duration.count(), 1000);  // Should process in less than 1 second
    
    EXPECT_EQ(tokens.size(), 40001);  // 10000 * 4 tokens per line + EOF
    for (size_t i = 0; i < tokens.size() - 1; i += 4) {
        expectToken(tokens[i], TokenType::TEXT, "x", (i / 4) + 1, 1);
        expectToken(tokens[i + 1], TokenType::ASSIGN, "=", (i / 4) + 1, 3);
        expectToken(tokens[i + 2], TokenType::NUMBER, "42", (i / 4) + 1, 5);
        expectToken(tokens[i + 3], TokenType::NEWLINE, "\n", (i / 4) + 1, 7);
    }
}

// Memory Usage Test
TEST_F(LexerTest, MemoryUsage) {
    std::string input;
    for (int i = 0; i < 1000; i++) {
        input += "\"This is a very long string that should be efficiently handled by the lexer without excessive memory allocation\"\n";
    }
    
    auto tokens = scan(input);
    EXPECT_EQ(tokens.size(), 2001);  // 1000 * 2 tokens per line (TEXT + NEWLINE) + EOF
    
    // Verify the first and last tokens
    expectToken(tokens[0], TokenType::TEXT, "\"This is a very long string that should be efficiently handled by the lexer without excessive memory allocation\"", 1, 1);
    expectToken(tokens[tokens.size() - 2], TokenType::NEWLINE, "\n", 1000, 96);
}

// Stress Test with Random Input
TEST_F(LexerTest, StressTest) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(32, 126);  // Printable ASCII characters
    
    std::string input;
    for (int i = 0; i < 10000; i++) {
        char c = static_cast<char>(dis(gen));
        input += c;
        if (i % 100 == 0) input += '\n';
    }
    
    auto tokens = scan(input);
    EXPECT_FALSE(Lexer(input).hadError());
}

// Error Recovery Test
TEST_F(LexerTest, ErrorRecovery) {
    auto tokens = scan(R"(
if x > 10 then
    y = "Hello
    z = 42
end
)");
    
    EXPECT_TRUE(Lexer(R"(
if x > 10 then
    y = "Hello
    z = 42
end
)").hadError());
    
    // Verify that lexing continues after the error
    expectToken(tokens[0], TokenType::IF, "if", 2, 1);
    expectToken(tokens[1], TokenType::TEXT, "x", 2, 4);
    expectToken(tokens[2], TokenType::GREATER, ">", 2, 6);
    expectToken(tokens[3], TokenType::NUMBER, "10", 2, 8);
    expectToken(tokens[4], TokenType::THEN, "then", 2, 11);
    expectToken(tokens[5], TokenType::NEWLINE, "\n", 2, 15);
    expectToken(tokens[6], TokenType::TEXT, "y", 3, 5);
    expectToken(tokens[7], TokenType::ASSIGN, "=", 3, 7);
    expectToken(tokens[8], TokenType::ERROR, "Unterminated string", 3, 9);
} 